flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/SpeculativeScheduler.java

```diff
@@ -0,0 +1,340 @@
+        final SpeculativeExecutionVertex executionVertex =
+                getExecutionVertex(execution.getVertex().getID());
+        final ExecutionAttemptID attemptId = execution.getAttemptId();
+
+        // when an execution fails, remove it from current executions to make room for future
+        // speculative executions
+        executionVertex.archiveFailedExecution(attemptId);
+        executionSlotAllocator.cancel(attemptId);
+
+        super.onTaskFailed(execution);
+    }
+
+    @Override
+    protected void handleTaskFailure(
+            final Execution failedExecution, @Nullable final Throwable error) {
+
+        final SpeculativeExecutionVertex executionVertex =
+                getExecutionVertex(failedExecution.getVertex().getID());
+
+        // if the execution vertex is not possible finish or a PartitionException occurred, trigger
+        // an execution vertex failover to recover
+        if (!isExecutionVertexPossibleToFinish(executionVertex)
+                || ExceptionUtils.findThrowable(error, PartitionException.class).isPresent()) {
+            super.handleTaskFailure(failedExecution, error);
+        } else {
+            // add the execution failure to exception history even though not restarting the entire
+            // execution vertex
+            final long timestamp = System.currentTimeMillis();
+            setGlobalFailureCause(error, timestamp);
+            final FailureHandlingResult failureHandlingResult =
+                    executionFailureHandler.getFailureHandlingResult(
+                            failedExecution, error, timestamp);
+            if (failureHandlingResult.canRestart()) {
+                archiveFromFailureHandlingResult(
+                        createFailureHandlingResultSnapshot(failureHandlingResult));
+            } else {
+                failJob(error, timestamp);
+            }
+        }
+    }
+
+    private static boolean isExecutionVertexPossibleToFinish(
+            final SpeculativeExecutionVertex executionVertex) {
+        boolean anyExecutionPossibleToFinish = false;
+        for (Execution execution : executionVertex.getCurrentExecutions()) {
+            // if any execution has finished, no execution of the same execution vertex should fail
+            // after that
+            checkState(execution.getState() != ExecutionState.FINISHED);
+
+            if (execution.getState() == ExecutionState.CREATED
+                    || execution.getState() == ExecutionState.SCHEDULED
+                    || execution.getState() == ExecutionState.DEPLOYING
+                    || execution.getState() == ExecutionState.INITIALIZING
+                    || execution.getState() == ExecutionState.RUNNING) {
+                anyExecutionPossibleToFinish = true;
+            }
+        }
+        return anyExecutionPossibleToFinish;
+    }
+
+    @Override
+    protected void cancelAllPendingSlotRequestsInternal() {
```

Comment:
`cancelAllPendingSlotRequestsInternal` `cancelAllPendingSlotRequestsForVertices` and `cancelAllPendingSlotRequests` is not needed.

Reply:
I was used in `DefaultScheduler` but was removed in the fix-up commit https://github.com/apache/flink/pull/20222/commits/5bc0c1220ef77f997e910b14a4ea006e8d7a1f29.